src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -1049,113 +1045,156 @@ public bool TakeStartLine(ReadableBuffer buffer, out ReadCursor consumed, out Re
+                var length = span.Length;
+                for (var i = 0; i < length; i++)
                 {
-                    RejectRequest(RequestRejectionReason.UnrecognizedHTTPVersion, httpVersion);
+                    var ch = data[i];
+
+                    switch (state)
+                    {
+                        case StartLineState.ExpectKnownMethod:
+                            if (span.GetKnownMethod(out method))
+                            {
+                                // Update the index, current char, state and jump directly
+                                // to the next state
+                                i += method.Length + 1;
+                                ch = data[i];
+                                state = StartLineState.ExpectPath;
+
+                                goto case StartLineState.ExpectPath;
+                            }
+
+                            state = StartLineState.ExpectUnknownMethod;
+                            goto case StartLineState.ExpectUnknownMethod;
+
+                        case StartLineState.ExpectUnknownMethod:
+                            if (ch == ByteSpace)
+                            {
+                                method = span.Slice(0, i).GetAsciiString();
+
+                                if (method == null)
+                                {
+                                    RejectRequestLine(start, end);
+                                }
+
+                                state = StartLineState.ExpectPath;
+                            }
+                            else if (!IsValidTokenChar((char)ch))
+                            {
+                                RejectRequestLine(start, end);
+                            }
+
+                            break;
+                        case StartLineState.ExpectPath:
+                            if (ch == ByteSpace)
+                            {
+                                pathEnd = i;
+
+                                if (pathStart == -1)
+                                {
+                                    // Empty path is illegal
+                                    RejectRequestLine(start, end);
+                                }
+
+                                // No query string found
+                                queryStart = queryEnd = i;
+
+                                state = StartLineState.ExpectKnownVersion;
+                            }
+                            else if (ch == ByteQuestionMark)
+                            {
+                                pathEnd = i;
+                                queryStart = i;
+                                state = StartLineState.ExpectQueryString;
+                            }
+                            else if (ch == BytePercentage)
+                            {
+                                needDecode = true;
+                            }
+
+                            if (pathStart == -1)
```

Comment:
Did you think about having another state to capture start and then switch to another to parse the rest?